### DISCUSSION TITLE
🐞 Failed run erases the cursor in a terminal

### DISCUSSION THREAD

**Discussion by esafak** [dagger/dagger#12312] [Help]
### What is the issue?

I ran my CI script `dagger run npx tsx ci.mts` and it made my block cursor disappear.

### Dagger version

dagger v0.11.0 ([registry.dagger.io/engine](http://registry.dagger.io/engine)) darwin/arm64

### Steps to reproduce

`dagger run npx tsx ci.mts` with some Docker stuff that fails.

### Log output

```
✘ Container.stderr: String! 47.2s
! failed to compute cache key: write /var/lib/dagger/runc-overlayfs/snapshots/snapshots/36/fs/usr/lib/aarch64-linux-gnu/libx265.so

        date: [ 'Mon, 22 Apr 2024 23:03:22 GMT' ],
        connection: [ 'close' ]
      }
    }
  }
}

Node.js v18.20.2
exit status 1
```

**Comment by cchanley2003** [comment]
So I have seen this on zsh on Ubuntu. Also it does something where stdin for future commands doesn't work. For instance if I have a failed dagger run and then I attempt to run a sudo command in the same terminal I can't type in my password. Or if a command requests confirmation that doesn't work. I haven't found a way to fix it other then close the terminal and start fresh.

The dagger command does replace the cursor, but stdin is still broken.

**Comment by mimizone** [comment]
it's easy to reproduce by just not starting Docker on macOS.

The same happens in a bash terminal. Using iTerm or Terminal.
Not only the cursor disappears in this case. But also the newline is messed up. Pressing just enter on the prompt line doesn't create a new line and just keeps adding the prompt again and again.
And there is no text echo, no text shows up when typing or copy pasting anything.
It also doesn't react to CTRL+C etc...


One way to get the cursor back is to open vi and exit.
On zsh everything is back to normal with that, but not the CTRL+C issue.
But on bash it doesn't solve any of the issues above.

**Comment by d3287t328** [comment]
I have seen this in bash on ubuntu. It seems to only happen when running a task fails due to a bug in the pipeline code.

**Comment by levlaz** [comment]
I also saw this yesterday in bash on ubuntu via a windows client - seems its not specific to OS or shell to me. I updated the title of this issue to make that part a bit more clear.

**Comment by wingyplus** [comment]
More info, the PowerShell is not broken. With reproducing code (Elixir):

```elixir
Mix.install([:dagger])

dag = Dagger.connect!()

{:ok, _} = dag
|> Dagger.Client.container()
|> Dagger.Container.from("alpine")
|> Dagger.Container.with_exec(~w"exit 1")
|> Dagger.Container.stdout()

Dagger.close(dag)
```

**Comment by easyas314159** [comment]
I am also losing my cursor and stdin on nixos+zsh. Based on my experiences with 0.11.4 and this bug over the last week, it looks like whatever terminal settings are being changed by the `--progress tty` flag are not restored at the end of any incomplete run. However, an incomplete run followed by a successful run does not restore the terminal settings. I have no issues when exclusively using `--progress plain`.

You can trigger the same tty behaviour if the build is interrupted partway through, e.g. after receiving a SIGINT. If you hit ctrl-C before the 10 second sleep finishes in the below example you'll also trigger the issue:

```go
package main

import (
	"context"
	"time"

	"dagger.io/dagger"
)

func main() {
	ctx := context.Background()

	client, err := dagger.Connect(ctx)
	if err != nil {
		return
	}

	client.Container().
		From("alpine:latest").
		WithEnvVariable("NOCACHE", time.Now().String()).
		WithExec([]string{"sleep", "10"}).
		Sync(ctx)
}
```

**Comment by cchanley2003** [comment]
Running 0.11.6 and I am not seeing the same issues with losing my command line. Maybe this can be closed?

**Comment by levlaz** [comment]
I am still seeing this in some cases, in particular when I am running a command with --debug mode on and it fails, I don't fully understand what causes this
